chat.go

```diff
@@ -0,0 +1,124 @@
+// Package chat implements a chat server & client
+package chat
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Server represents a chat server
+type Server struct {
+	Logger         logrus.StdLogger // with standard logger can be extended with logrus
+	WelcomeMessage string
+	listener       net.Listener
+	clients        map[*net.Conn]bool
+	register       chan *net.Conn
+	unregister     chan *net.Conn
+	broadcast      chan string
+}
+
+// Logger enables a customization of the log function
+type Logger interface {
+	Println(v ...interface{})
+}
+
+// Run implements the logic handling connections
+func (s *Server) Run() {
+
+	go s.listen()
+
+	for {
+		select {
+		case conn := <-s.register:
+			s.clients[conn] = true
+			s.Logger.Println("client connection registered")
+			go s.handle(conn)
+		case m := <-s.broadcast:
```

Comment:
Let's make this `msg`; it's a lot clearer for only a couple of extra characters.